File: xdl/test/python/unit_test/backend/mxnet_backend_test.py
# Copyright 2018 Alibaba Group. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import xdl
import unittest
import numpy as np
from xdl.python.lib.datatype import *
from xdl.python.lib.graph import execute
from xdl.python.backend.mxnet.mxnet_backend import *

def main():
  dense = xdl.mock_dense_op(shape=[1, 16], value=0.01, name_="dense")
  gear = xdl.mock_dense_op(shape=[1, 1], value=0.01, name_="gear")
  labels = xdl.mock_dense_op(shape=[1, 1], value=1.0, name_="label")
  gear.set_shape([1, 1])
  dense.set_shape([1, 16])
  labels.set_shape([1, 1])
  with xdl.model_scope("ams_main"):
    loss = ams_main(main_model)(dense, labels, gear_inputs=[gear])
    sess = xdl.TrainSession()
    return sess.run([xdl.get_collection("gear_grad")])

def gear():
  forward = xdl.mock_dense_op(shape=[1, 16], value=0.01, name_="forward")
  backward = xdl.mock_dense_op(shape=[1, 16], value=0.02, name_="backward")
  labels = xdl.mock_dense_op(shape=[1, 1], value=1.0, name_="label1")
  init_grad = xdl.mock_dense_op(shape=[1, 1], value=0.3, name_="init_grad")
  forward.set_shape([1, 16])
  backward.set_shape([1, 16])
  labels.set_shape([1, 1])  
  init_grad.set_shape([1, 1])  
  predict = ams_gear([forward], [backward], init_grad)(gear_model)(None)
  with xdl.model_scope("ams_gear_forward"):
    sess = xdl.TrainSession()
    prediction = sess.run(predict)
  with xdl.model_scope("ams_gear_backward"):
    grads = xdl.get_gradient("fc_weight")
    sess = xdl.TrainSession()
    fc_weight_grad = sess.run(grads)
    return prediction, fc_weight_grad

def main_model(dense, label, gear_inputs):
  weight = mx.sym.var(name='fc_weight', init=mx.init.Constant(0.1))
  fc = mx.symbol.FullyConnected(data=dense, num_hidden=1, weight=weight, name="fc")
  logits = fc + gear_inputs[0]
  return mx.sym.MakeLoss(logits)

def gear_model(inputs):
  weight = mx.sym.var(name='fc_weight', init=mx.init.Constant(0.1))
  fc = mx.sym.FullyConnected(data=inputs, num_hidden=1, name='fc', weight=weight)
  return fc

class MxnetBackendTest(unittest.TestCase):
  def test_ams_main(self):
    gear_gradient = main()
    self.assertTrue(gear_gradient[0]==np.array([[1.0]], dtype=np.float32))

  def test_ams_gear(self):
    prediction, grad = gear()
    self.assertTrue((prediction==np.array([[0.016]], dtype=np.float32)).all())
    self.assertTrue((grad==np.repeat(np.array([0.006], dtype=np.float32), 16).reshape(1,16)).all())

def suite():
  return unittest.TestLoader().loadTestsFromTestCase(MxnetBackendTest)

if __name__ == '__main__':
  unittest.TextTestRunner().run(suite())

File: xdl/test/core/backend/mxnet/mxnet_backend_test.cc
/* Copyright (C) 2016-2018 Alibaba Group Holding Limited

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "gtest/gtest.h"
#include "xdl/core/framework/tensor.h"
#include "xdl/core/utils/file_utils.h"
#include "xdl/core/backend/device_singleton.h"
#include "xdl/core/backend/mxnet/mxnet_runner.h"
#include "xdl/core/backend/mxnet/convert_utils.h"

#include <stdlib.h>

using namespace xdl;

static void ArrayEqual(const mxnet::cpp::NDArray& src,
                       const std::vector<float>& dst) {
  ASSERT_EQ(src.Size(), dst.size());
  const float* src_ptr = src.GetData();
  for (size_t i = 0; i < dst.size(); ++i) {
    ASSERT_FLOAT_EQ(src_ptr[i], dst[i]);
  }
}

TEST(MxnetBackendTest, RunnerTest) {
  MxnetRunner runner;
  std::string symbol_path = "../test_data/backend/mx_test.json";
  std::string graph_def = FileUtils::ReadLocalFile(symbol_path);
  ASSERT_TRUE(!graph_def.empty());
  ASSERT_TRUE(Status::Ok() == runner.Init(graph_def, "cpu"));

  for (size_t n = 0; n < 3; ++n) {
    size_t bs = 2 + rand() / (RAND_MAX / 8);
    printf("==== bs=%lu ====\n", bs);

    Tensor a(DeviceSingleton::CpuInstance(), TensorShape({bs}), DataType::kFloat);
    Tensor b(DeviceSingleton::CpuInstance(), TensorShape({bs}), DataType::kFloat);
    Tensor y(DeviceSingleton::CpuInstance(), TensorShape({bs}), DataType::kFloat);
    mx_float* a_ptr = new mx_float[bs];
    mx_float* b_ptr = new mx_float[bs];
    mx_float* y_ptr = new mx_float[bs];
    for (int i = 0; i < bs; ++i) {
      a_ptr[i] = 1;
      b_ptr[i] = 2;
      y_ptr[i] = 2 + i;
    }
    memcpy(a.Raw<float>(), a_ptr, bs * sizeof(float));
    memcpy(b.Raw<float>(), b_ptr, bs * sizeof(float));
    memcpy(y.Raw<float>(), y_ptr, bs * sizeof(float));

    MxnetRunner::InputList inputs;
    inputs.push_back({"a", a});
    inputs.push_back({"b", b});
    inputs.push_back({"y", y});

    MxnetRunner::DataList outputs;
    MxnetRunner::DataList gradients;
  
    ASSERT_TRUE(Status::Ok() ==
                runner.Run(inputs, &outputs, &gradients));

    ASSERT_EQ(2, outputs.size());
    ASSERT_EQ(outputs[0].Size(), bs);
    ASSERT_EQ(outputs[1].Size(), bs);
    for (size_t i = 0; i < bs; ++i) {
      ASSERT_FLOAT_EQ(outputs[0].At(0, i), -0.5 + 0.5 * i);
      ASSERT_FLOAT_EQ(outputs[1].At(0, i), 3);
    }
    //ArrayEqual(outputs[0], {-0.5, 0});  // bs = 2
    //ArrayEqual(outputs[1], {3, 3});

    ASSERT_EQ(3, gradients.size());
    ASSERT_EQ(gradients[0].Size(), bs);
    ASSERT_EQ(gradients[1].Size(), bs);
    ASSERT_EQ(gradients[2].Size(), bs);
    for (size_t i = 0; i < bs; ++i) {
      ASSERT_FLOAT_EQ(gradients[0].At(0, i), -0.5);
      ASSERT_FLOAT_EQ(gradients[1].At(0, i), -0.5);
      ASSERT_FLOAT_EQ(gradients[2].At(0, i), 0.5);
    }
    //ArrayEqual(gradients[0], {-0.5, -0.5});  // bs = 2
    //ArrayEqual(gradients[1], {-0.5, -0.5});
    //ArrayEqual(gradients[2], {0.5, 0.5});

    delete[] a_ptr;
    delete[] b_ptr;
    delete[] y_ptr;
  }
}

